### PR TITLE
Dynamic groups

### PR DESCRIPTION
**Pull request by sashankaryal**
## What changes are proposed in this pull request?

- Dynamic groups app support (backend tracked by #2475)
- Group code modularization and some cleanup

dynamic_groups-small

Notes to the reviewer:
- Pay special attention to files in the `app/packages/core/src/components/Modal/Group/DynamicGroup` directory

## How is this patch tested? If it is not, please explain why.

Dataset for testing was generated using the following snippet. Every th samples are folded into a group (here, 25).
```
import fiftyone as fo
import fiftyone.zoo as foz
from fiftyone import ViewField as F

dynamic_groups = foz.load_zoo_dataset("quickstart-groups").clone()

every_th_sample = 25
# for each slice, for every th samples, assign an arbitrary "scene_id"
for slice in dynamic_groups.group_slices:
    dynamic_groups.group_slice = slice
    scene_id_counter = 0
    order_by_counter = 0
    for sample in dynamic_groups:
        sample.set_field("scene_id", scene_id_counter // every_th_sample)
        sample.set_field("timestamp", order_by_counter % every_th_sample)
        sample.save()
        scene_id_counter = scene_id_counter + 1
        order_by_counter = order_by_counter + 1
dynamic_groups.save()
```

### Is this a user-facing change that should be mentioned in the release notes?

-   [ ] No. You can skip the rest of this section.
-   [x] Yes. Give a description of this change to be included in the release
        notes for FiftyOne users.

(Details in 1-2 sentences. You can just refer to another PR with a description
if this PR is part of a larger change.)

### What areas of FiftyOne does this PR affect?

-   [x] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [x] Core: Core `fiftyone` Python library changes
-   [x] Documentation: FiftyOne documentation changes
-   [ ] Other

### CODE REVIEW

File: fiftyone/server/samples.py [hunk 1]
```diff
@@ -22,6 +22,16 @@
...
 
+import logging
+
+logging.basicConfig(
+    filename="/tmp/fiftyone.log",
+    filemode="a",
+    format="%(asctime)s,%(msecs)d %(name)s %(levelname)s %(message)s",
+    datefmt="%H:%M:%S",
+    level=logging.INFO,
+)
```
Comment (benjaminpkane) [inline fiftyone/server/samples.py:33]: Is this for debugging? I'm not sure this should be used in production. Should probably discuss with other backend folks

File: app/packages/core/src/components/Actions/DynamicGroupAction.tsx [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+  const [useOrdered, setUseOrdered] = useState(false);
+
+  const ref = useRef<HTMLDivElement>(null);
+  const [measureRef] = useMeasure();
+
+  const onComplete = useRecoilCallback(({ set, reset }) => () => {
+    setIsFieldValidated(false);
+    setIsProcessing(false);
+  });
```
[the remark this answers is not in the document]
Comment (sashankaryal) [reply app/packages/core/src/components/Actions/DynamicGroupAction.tsx:46]: oops sorry

File: fiftyone/server/samples.py (hunk 1 above)
Comment (sashankaryal) [reply fiftyone/server/samples.py:33]: ah, this is just for debugging, didn't mean to commit.

File: app/packages/core/src/components/ImageContainerHeader.tsx [hunk 3]
```diff
@@ -109,9 +110,15 @@ const ImageContainerHeader = () => {
   const setGridZoom = useSetRecoilState(gridZoom);
   const gridZoomRangeValue = useRecoilValue(gridZoomRange);
   const theme = useTheme();
-  const group = useRecoilValue(isGroup);
+  const isGroup = useRecoilValue(isGroupAtom);
+  const groupSlices = useRecoilValue(fos.groupSlices);
   const groupStats = useRecoilValue(groupStatistics(false));
 
+  const shouldShowSliceSelector = useMemo(
+    () => isGroup && groupSlices.length > 0,
```
Comment (benjaminpkane) [inline app/packages/core/src/components/ImageContainerHeader.tsx:118]: Could be wrong, but should it be `> 1`?

**Review by benjaminpkane — APPROVED** [review]
LGTM!

**Review by benjaminpkane — APPROVED** [review]
LGTM!

**Comment by sashankaryal** [comment]
> LGTM!

Thanks for leaving the comments. Will address comments in another PR.

File: app/packages/core/src/components/Modal/Group/GroupView.tsx [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+              <GroupSample3d />
+            ) : null}
+          </Resizable>
+        )}
+
+        {shouldSplitVertically && <GroupSample3d />}
+
+        {!shouldSplitVertically && is3DVisible && <GroupSample3d />}
+      </div>
+      {subBar ?? subBar}
```
[the remark this answers is not in the document]
Comment (sashankaryal) [reply app/packages/core/src/components/Modal/Group/GroupView.tsx:108]: Meant to do `&&`

File: app/packages/flashlight/src/index.ts [hunk 5]
```diff
@@ -56,11 +64,32 @@ export default class Flashlight<K> {
     this.container = this.createContainer();
     this.showPixels();
     this.element = document.createElement("div");
+    config.elementId && this.element.setAttribute("id", config.elementId);
     this.element.classList.add(flashlight);
     this.state = this.getEmptyState(config);
 
     document.addEventListener("visibilitychange", () => this.render());
 
+    if (config.enableKeyNavigation) {
```
[the remark this answers is not in the document]
Comment (sashankaryal) [reply app/packages/flashlight/src/index.ts:73]: I'll add a guard for horizontal nav

**Pull request merged**